File: lib/common/pnfft/tests/pnfft_test.c
#include <stdlib.h>
#include <complex.h>
#include <pnfft.h>

static void init_random_x(
    const double *lo, const double *up, ptrdiff_t local_M,
    double *x);
static double random_number_less_than_one(
    void);
static void vpr_complex(
    MPI_Comm comm, ptrdiff_t num, pnfft_complex *vec, const char *info);


int main(int argc, char **argv){
  int np[3];
  ptrdiff_t N[3], local_M;
  ptrdiff_t local_N[3], local_N_start[3];
  double lower_border[3], upper_border[3];
  MPI_Comm comm_cart_3d;
  pnfft_complex *f_hat, *f;
  double *x;
  pnfft_plan pnfft;
  
  MPI_Init(&argc, &argv);
  pnfft_init();
  
  /* Set default values */
  N[0] = N[1] = N[2] = 16;
  np[0]=2; np[1]=2; np[2]=2;
  local_M = N[0]*N[1]*N[2]/(np[0]*np[1]*np[2]);
  
  /* Print infos */
  pfft_printf(MPI_COMM_WORLD, "******************************************************************************************************\n");
  pfft_printf(MPI_COMM_WORLD, "* Computation of parallel NFFT\n");
  pfft_printf(MPI_COMM_WORLD, "* for  N[0] x N[1] x N[2] = %td x %td x %td Fourier coefficients\n", N[0], N[1], N[2]);
  pfft_printf(MPI_COMM_WORLD, "* at   local_M = %td nodes per process\n", local_M);
  pfft_printf(MPI_COMM_WORLD, "* on   np[0] x np[1] x np[2] = %td x %td x %td processes\n", np[0], np[1], np[2]);
  pfft_printf(MPI_COMM_WORLD, "*******************************************************************************************************\n\n");

  /* create three-dimensional process grid of size np[0] x np[1] x np[2], if possible */
  if( pnfft_create_procmesh(3, MPI_COMM_WORLD, np, &comm_cart_3d) ){
    pfft_fprintf(MPI_COMM_WORLD, stderr, "Error: Procmesh of size %d x %d x %d does not fit to number of allocated processes.\n", np[0], np[1], np[2]);
    pfft_fprintf(MPI_COMM_WORLD, stderr, "       Please allocate %d processes (mpiexec -np %d ...) or change the procmesh (with -pnfft_np * * *).\n", np[0]*np[1]*np[2], np[0]*np[1]*np[2]);
    MPI_Finalize();
    return 1;
  }

  /* Get parameters of data distribution */
  pnfft_local_size_3d(N, comm_cart_3d,
      local_N, local_N_start, lower_border, upper_border);

  /* Plan parallel NFFT */
  pnfft = pnfft_init_3d(N, local_M, comm_cart_3d);

  /* get data pointers */
  f_hat = pnfft_get_f_hat(pnfft);
  f     = pnfft_get_f(pnfft);
  x     = pnfft_get_x(pnfft);

  /* Initialize Fourier coefficients */
  pnfft_init_f_hat_3d(N, local_N, local_N_start,
      f_hat);

  /* Initialize nonequispaced nodes */
  init_random_x(lower_border, upper_border, local_M,
      x);

  /* Print input Fourier coefficents */
  vpr_complex(comm_cart_3d, 8, f_hat,
      "Input Fourier coefficients on process 1:");

  /* Execute parallel NFFT */
  pnfft_trafo(pnfft);

  /* Print NFFT results */
  vpr_complex(comm_cart_3d, 8, f,
      "PNFFT Results on process 1:");

  /* Execute parallel adjoint NFFT */
  pnfft_adj(pnfft);

  /* Scale data */
  for(ptrdiff_t l=0; l < local_N[0] * local_N[1] * local_N[2]; l++)
    f_hat[l] /= (N[0]*N[1]*N[2]);

  /* Print output Fourier coefficents */
  vpr_complex(comm_cart_3d, 8, f_hat,
      "Fourier coefficients after one forward and backward PNFFT on process 1:");

  /* free mem and finalize */
  pnfft_finalize(pnfft, PNFFT_FREE_X | PNFFT_FREE_F_HAT| PNFFT_FREE_F);
  MPI_Comm_free(&comm_cart_3d);
  pnfft_cleanup();
  MPI_Finalize();
  return 0;
}


static void vpr_complex(
    MPI_Comm comm, ptrdiff_t num, pnfft_complex *vec, const char *info
    )
{
  int myrank;
  MPI_Comm_rank(comm, &myrank);

  if(myrank==0){
    printf("%s\n", info);
    for(ptrdiff_t k=0; k<num; k++){
      if(k%4 == 0)
        printf("%4td. ", k);
      printf("%.2e + %.2e I\t", creal(vec[k]), cimag(vec[k]));
      if(k%4 == 3)
        printf("\n");
    }
  }
}


static void init_random_x(
    const double *lo, const double *up, ptrdiff_t local_M,
    double *x
    )
{
  double tmp;
  
  for (ptrdiff_t j=0; j<local_M; j++){
    for(int t=0; t<3; t++){
      do{
        tmp = random_number_less_than_one();
        tmp = (up[t]-lo[t]) * tmp + lo[t];
      }
      while( (tmp < -0.5) || (0.5 <= tmp) );
      x[3*j+t] = tmp;
    }
  }
}


static double random_number_less_than_one(
    void
    )
{
  double tmp;
  
  do
    tmp = ( 1.0 * rand()) / RAND_MAX;
  while(tmp>=1.0);
  
  return tmp;
}

File: lib/common/pnfft/tests/f03/pnfft_test.f90

program main

  use, intrinsic :: iso_c_binding
  implicit none
  include "mpif.h"
  include "fftw3-mpi.f03"
  include "pfft.f03"
  include "pnfft.f03"

  integer np(3), ierror
  integer(C_INTPTR_T) l1, l2, l3
  integer(C_INTPTR_T) :: N(3), local_N(3), local_N_start(3)
  integer(C_INTPTR_T) :: local_M, d=3
  real(C_DOUBLE) :: lower_border(3), upper_border(3)
  type(C_PTR) :: pnfft, cf_hat, cf, cx
  complex(C_DOUBLE_COMPLEX), pointer :: f_hat(:,:,:), f(:)
  real(C_DOUBLE), pointer :: x(:,:)

  integer comm_cart_3d, myrank

  N = (/ 16,16,16 /)
  np = (/ 2,2,2 /)
  local_M = N(1)*N(2)*N(3) / (np(1)*np(2)*np(3))

! Initialize MPI and PFFT
  call MPI_Init(ierror)
  call pnfft_init();
  call MPI_Comm_rank(MPI_COMM_WORLD, myrank, ierror)
  
! Create three-dimensional process grid of
! size np(1) x np(2) x np(3), if possible
  ierror =  pnfft_create_procmesh(3, MPI_COMM_WORLD, np, comm_cart_3d)
  if (ierror .ne. 0) then
    if(myrank .eq. 0) then
      write(*,*) "Error: This test file only works with ", np(1)*np(2)*np(3), " processes"
    endif
    call MPI_Finalize(ierror)
    call exit(1)
  endif

! Get parameters of data distribution
  call pnfft_local_size_3d(N, comm_cart_3d, PNFFT_TRANSPOSED_NONE, &
      local_N, local_N_start, lower_border, upper_border)

! Plan parallel NFFT
  pnfft = pnfft_init_3d(N, local_M, comm_cart_3d)

! Get data pointers in C format
  cf_hat = pnfft_get_f_hat(pnfft)
  cf     = pnfft_get_f(pnfft)
  cx     = pnfft_get_x(pnfft)

! Convert data pointers to Fortran format
  call c_f_pointer(cf_hat, f_hat, [local_N])
  call c_f_pointer(cf,     f,     [local_M])
  call c_f_pointer(cx,     x,     [d,local_M])

! Initialize Fourier coefficients
  call init_f_hat(N, local_N, local_N_start, &
      f_hat)

! Initialize nonequispaced nodes
  call init_random_x(lower_border, upper_border, local_M, &
     x)

! Print input Fourier coefficents
  if(myrank .eq. 0) then
    write (*,*) "Input Fourier coefficients on process 1:"
    write (*,*) f_hat(1:4,1,1)
  endif

! Execute parallel NFFT
  call pnfft_trafo(pnfft)

! Print NFFT results
  if(myrank .eq. 0) then
    write (*,*) ""
    write (*,*) "PNFFT Results on process 1:"
    write (*,*) f(1:4)
  endif

! Execute parallel adjoint NFFT
  call pnfft_adj(pnfft)

! Scale data
  do l3=1,local_N(3)
    do l2=1,local_N(2)
      do l1=1,local_N(1)
        f_hat(l1,l2,l3) = f_hat(l1,l2,l3) / (N(3)*N(2)*N(1))
      enddo
    enddo
  enddo

! Print output Fourier coefficents
  if(myrank .eq. 0) then
    write (*,*) ""
    write (*,*) "Fourier coefficients after one forward and backward PNFFT on process 1:"
    write (*,*) f_hat(1:4,1,1)
  endif

! Free mem and finalize      
  call pnfft_finalize(pnfft, &
      PNFFT_FREE_X + PNFFT_FREE_F_HAT + PNFFT_FREE_F)
  call MPI_Comm_free(comm_cart_3d, ierror)
  call pnfft_cleanup()
  call MPI_Finalize(ierror)
end program main


subroutine init_random_x(lo, up, local_M, x)
  use, intrinsic :: iso_c_binding
  implicit none

  integer(C_INTPTR_T) j, t
  integer(C_INTPTR_T), intent(in) :: local_M
  real(C_DOUBLE), intent(in) :: lo(3), up(3)
  real(C_DOUBLE), intent(out) :: x(3,local_M)

  do j=1,local_M
    do t=1,3
      x(t,j) = (up(t) - lo(t)) * rand(0) + lo(t)
    enddo
  enddo
end subroutine init_random_x

subroutine init_f_hat(N, local_N, local_N_start, f_hat)
  use, intrinsic :: iso_c_binding
  implicit none

  integer(C_INTPTR_T) :: l1, l2, l3
  integer(C_INTPTR_T), intent(in) :: N(3), local_N(3), local_N_start(3)
  complex(C_DOUBLE_COMPLEX), intent(out) :: f_hat(local_N(1),local_N(2),local_N(3))
  real :: func

! use C-like row-major order here  
  do l1 = 1,local_N(1)
    do l2 = 1,local_N(2)
      do l3 = 1,local_N(3)
        f_hat(l1,l2,l3) = &
            func(l1 + local_N_start(1), N(1)) & 
          * func(l2 + local_N_start(2), N(2)) &
          * func(l3 + local_N_start(3), N(3)) 
      enddo
    enddo
  enddo
end subroutine init_f_hat

real function func(l,N)
  use, intrinsic :: iso_c_binding
  implicit none

  integer(C_INTPTR_T), intent(in) :: l, N

  func = exp(-real(l*l)/real(N*N))
end function func

